### ISSUE TITLE
Question regarding dynamic buffers

### ISSUE THREAD

**Issue by MikeHart66** [floooh/sokol#449]
As there is no community where i could ask such a simple question...
Which example shows the use of dynamic buffers the best?
So far I got the notion that everything is predefined.

**Comment by floooh** [comment]
The best place to look for feature examples is the sokol-samples page and click the source code link:

https://floooh.github.io/sokol-html5/

For instance the instancing-sapp sample uses a dynamic vertex buffer to update the particle positions:

https://floooh.github.io/sokol-html5/instancing-sapp.html

Some of the util headers also use dynamic vertex- and index-buffers (look for sg_make_buffer(), sg_update_buffer() and sg_append_buffer()):

- https://github.com/floooh/sokol/blob/master/util/sokol_imgui.h
- https://github.com/floooh/sokol/blob/master/util/sokol_gl.h
- https://github.com/floooh/sokol/blob/master/util/sokol_debugtext.h

**Comment by MikeHart66** [comment]
Thanks Floooh for answering. And creating this awesome lib.
I wish there would be at least a discord channel or the new community feature of GH to ask these question but if you prefer it here, np.
If others can help out, feel free to jump in. I appreciate all info that I can get.

So my problem is that I hit a road block with porting my cross-platform game development language from using OpenGL to Sokol. Of course, Metal is the key reason here.

Color me stupid, I am not sure if i fully understood the buffer appending/updating/drawing concept.

During a frame, a user could do something like this (it is a kinda BASIC language, https://www.cerberus-x.com if you are interested).

`DrawImage( image1, 10,10) 
DrawPoint(200, 100)
DrawImage( image2, 200,200)
DrawImage( image2, 200,300)
DrawLine(0, 0, 300, 200)
DrawImage( image1, 300,10)
DrawImage( image2, 300,300)
DrawPoint(400, 100)`

All these drawing commands fill an array with informations. For each vertex  we fill coordinates, uv coordinates and color in a specific order. If an image (surface) is involved, then that is bound. 

At the begin of the drawing operation , I check if the type (point, line, image) aka primType  is different from the last one, if the image has changed or the max vertex count of the array is reached. If no changes are detected, the vertex offset is calculated and in the actual draw command the array is filled with the info.

`float *gxtkGraphics::Begin( int type,int count,gxtkSurface *surf ){
	if( primType!=type || primSurf!=surf || vertCount+count>MAX_VERTS ){
		Flush();
		primType=type;
		primSurf=surf;
	}
	float *vp=vertices+vertCount*5;
	vertCount+=count;
	return vp;
}`

If a Flush is need, I do the actual drawing...

`void gxtkGraphics::Flush(){
	if( !vertCount ) return;

	if( primSurf ){
		glEnable( GL_TEXTURE_2D );
		primSurf->Bind();
	}
		
	switch( primType ){
	case 1:
		glDrawArrays( GL_POINTS,0,vertCount );
		break;
	case 2:
		glDrawArrays( GL_LINES,0,vertCount );
		break;
	case 3:
		glDrawArrays( GL_TRIANGLES,0,vertCount );
		break;
	case 4:
		glDrawElements( GL_TRIANGLES,vertCount/4*6,GL_UNSIGNED_SHORT,quadIndices );
		break;
	default:
		for( int j=0;j<vertCount;j+=primType ){
			glDrawArrays( GL_TRIANGLE_FAN,j,primType );
		}
		break;
	}

	if( primSurf ){
		glDisable( GL_TEXTURE_2D );
	}

	vertCount=0;
}`

We store all the data in one array, just dont fill in some info for all primtTypes.

For an example a point 

`int gxtkGraphics::DrawPoint( float x,float y ){

	if( tformed ){
		float px=x;
		x=px * ix + y * jx + tx;
		y=px * iy + y * jy + ty;
	}
	
	float *vp=Begin( 1,1,0 );
	
	vp[0]=x+.5f;
	vp[1]=y+.5f;
	(int&)vp[4]=colorARGB;

	return 0;	
}`

or an image...

`int gxtkGraphics::DrawSurface( gxtkSurface *surf,float x,float y ){
	
	float w=surf->Width();
	float h=surf->Height();
	float x0=x,x1=x+w,x2=x+w,x3=x;
	float y0=y,y1=y,y2=y+h,y3=y+h;
	float u0=0,u1=w*surf->uscale;
	float v0=0,v1=h*surf->vscale;

	if( tformed ){
		float tx0=x0,tx1=x1,tx2=x2,tx3=x3;
		x0=tx0 * ix + y0 * jx + tx;y0=tx0 * iy + y0 * jy + ty;
		x1=tx1 * ix + y1 * jx + tx;y1=tx1 * iy + y1 * jy + ty;
		x2=tx2 * ix + y2 * jx + tx;y2=tx2 * iy + y2 * jy + ty;
		x3=tx3 * ix + y3 * jx + tx;y3=tx3 * iy + y3 * jy + ty;
	}
	
	float *vp=Begin( 4,4,surf );
	
	vp[0 ]=x0; vp[1 ]=y0; vp[2 ]=u0; vp[3 ]=v0; (int&)vp[4 ]=colorARGB;
	vp[5 ]=x1; vp[6 ]=y1; vp[7 ]=u1; vp[8 ]=v0; (int&)vp[9 ]=colorARGB;
	vp[10]=x2; vp[11]=y2; vp[12]=u1; vp[13]=v1; (int&)vp[14]=colorARGB;
	vp[15]=x3; vp[16]=y3; vp[17]=u0; vp[18]=v1; (int&)vp[19]=colorARGB;
	
	return 0;
}`

So depending on in which order the user draws on the screen, between drawing the same primType and Image there could be other images and primTypes involved. 

I understood that I can update a buffer only once during a frame but I can append to it multiple times. I would then append to a buffer as long nothing changes and if it does change, draw the content from a given index.

What would be the best approach for porting this? One Buffer for all primTypes together, or several buffer types for each primType? 

The parameter description of sg_draw confuses me, especially the base_element one. You wrote that for instancing it is the first vertex to draw from. So it is an index in the vertexbuffer???? 

But then you have an offset parameter 
`bindings.vertex_buffer_offsets[0] = offset;`

With non instance drawing we should have the base_element untouched.. means 0 ?

My questions...

Now I understood that the first parameter of sg_draw (base_element) is kinda the index of the buffer array. So by updating the index to a new start point, I could later draw elements of the appended vertex buffer?

Do I have to set the base_element for non instanced drawing always to 0 or coudl I use the vertex_buffer_offset instead?

With instance drawing, could I also use the vertex_buffer_offset and let the base_element set to 0 and set num_instances to the amount  of images/points/lines to be drawn?

Thank you for taking your time to read this lengthy text, I hope I made some sense and you guys can help a Sokol-Beginner out.

Best wishes
Michael

**Comment by MikeHart66** [comment]
Damn, can't get it right to insert code. Sorry.

**Comment by floooh** [comment]
Regarding the sg_draw() arguments question vs vertex buffer offsets:

First ```sg_draw(base_element, num_elements, num_instances)```:

If you're rendering only from a vertex buffer, than base_element and num_elements are vertex indices into the vertex buffer.

If you're rendering from a vertex buffer and index buffer, than base_element and num_elements are "index indices" into the index buffer instead.

If you're rendering only from a vertex buffer (no index buffer), than base_element and sg_bindings.vertex_buffer_offset are essentially the same thing (except that vertex_buffer_offset is in number of bytes, and base_element is in number of vertices).

But when rendering with an index buffer, there's an important difference:

The indices in the index buffer use the sg_bindings.vertex_buffer_offset as base. If the vertex_buffer_offset is 0, then an index=0 in the index buffer points to the beginning of the vertex buffer. If sg_bindings.vertex_buffer_offset is >0, then an index=0 in the index buffer points to vertex data starting at vertex_buffer_offset.

TL;DR: the sg_draw base_element is useful for rendering 'submeshes' (e.g. different materials on a mesh), while sg_bindings.vertex_buffer_offset is useful for stuffing entirely different geometries into the same vertex- and index-buffers.

Check out this sample for vertex_buffer_offset where a triangle and a quad are stuffed into the same vertex- and index-buffer, note how the indices for the quad start at zero:

https://github.com/floooh/sokol-samples/blob/f773129bde30bc8197934f66979283c17f6bf0c9/sapp/bufferoffsets-sapp.c#L35-L51

Without using sg_bindings.vertex_buffer_offset to "separate" the triangle- from the quad-vertex-data, the quad-indices would actually refer to triangle indices.

This same scenario is why OpenGL in later versions added the ```glDrawElementsBaseVertex()``` call: https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glDrawElementsBaseVertex.xhtml (the base vertex index has the same job as the vertex_buffer_offset). But since such 'base-vertex' draw calls are not supported in all backend 3D APIs, I had to use the vertex-buffer-offset method to achieve the same thing.

**Comment by floooh** [comment]
> Now I understood that the first parameter of sg_draw (base_element) is kinda the index of the buffer array. So by updating the index to a new start point, I could later draw elements of the appended vertex buffer?

Yes.

> Do I have to set the base_element for non instanced drawing always to 0 or coudl I use the vertex_buffer_offset instead?

Both may makes sense in different situations, see my explanation above about rendering with index buffer or without. Using base_element should be preferred since that's a bit more efficient.

> With instance drawing, could I also use the vertex_buffer_offset and let the base_element set to 0 and set num_instances to the amount of images/points/lines to be drawn?

num_instances isn't directly related to base_element versus vertex_buffer_offset, for your use case, instanced rendering probably doesn't make sense (it's only useful when rendering a high number of identical geometry). Instanced rendering requires a very different buffer setup (you need at least two vertex buffers, one with the static geometry that should be rendered many times, and another buffer with per-instance data... this is an entirely different topic though, googling for "hardware instancing" gives some good articles about the topic)

**Comment by MikeHart66** [comment]
Thanks for the great explanation. So far I have the port working. I went for the append_buffer route. 
Sadly I am not able to draw more than 300-400 images (means 300-400x append_buffer with 4 vertices (x,y,z,r,g,b,a,u,v) 
before Sokol tells me that the buffer is not big enough.

This basically the drawing code...
```cpp
	sg_apply_pipeline(pip);
	const int num_bytes = sizeof(float)*4*9;
	for (int i = 0; i < primCount; i++) {
		const void* data = vertices+(4*9*i);
		int offset = sg_append_buffer(vbuf, data, num_bytes);
		sokbind.vertex_buffer_offsets[0] = offset;
    	        sokbind.fs_images[0] = primSurf->image;	
		sg_apply_bindings(&sokbind);
		sg_apply_uniforms(SG_SHADERSTAGE_VS, 0, &vs_params, sizeof(vs_params));
		sg_draw(0,6,1);
	}
```

I can't go the sokol_GL route as I need shaders down the road and to me it looks like OpenGL 1.2 only functionality.

Any idea?

**Comment by floooh** [comment]
How big is vbuf (in the sg_make_buffer() call)? There shouldn't be a limitation of the number of sg_append_buffer() calls, but of course the buffer can run full if the size is too small.

Note that this way drawing quads will be extremely slow though.

> I can't go the Sokol_GL route as I need shaders down the road

There's actually a somewhat "hidden" feature in sokol_gl.h to use your own shader, sokol_fontstash.h uses this. When creating a sokol-gl pipeline object, you can set a shader in the sg_pipeline_desc which will then be used by sokol-gl:

https://github.com/floooh/sokol/blob/924f13b8715fed762d8bed9ae3b043e89722c270/util/sokol_fontstash.h#L1661-L1670

The shader must be compatible with sokol-gl though, which still makes it fairly inflexible (that's why I didn't document this "feature"):

https://github.com/floooh/sokol/blob/924f13b8715fed762d8bed9ae3b043e89722c270/util/sokol_fontstash.h#L1594-L1659

**Comment by MikeHart66** [comment]
> How big is vbuf (in the sg_make_buffer() call)? 
Currently 1024*9vertices*4byte. I noticed already that I could make it bigger. The problem I see here is that i might have an example where Sokol doesn't fit in. Like I said, I am drawing 2 different images 50000x randomly over the screen during a frame.
The positions are defined by vertices of each corner. I know 50000x is  overkill but I try to see if Sokol is up to the task.
As they are drawn after another (img1, img2, img1, img2, etc.), instancing is not possible. Currently the only limitation in our current OpenGL implementation is the computer you run it on. Because the buffer is reused right now. 
In Sokol I can only define it at the beginning and then update it once during a frame. So defining a buffer that huge would not be wise, something tells me.

> Note that this way of drawing quads will be extremely slow though, because issuing one draw call per quad is slow on any 3D API (but especially slow on GL).
Sure. I was looking through your examples. Almost all have a buffers of fixed size, with fixed vertex coordinates. Or you go the append buffer route but with a much smaller amount of images. The only example with instancing is the particle one, but that is the same object, and it doesn't fit here.

My last question.... why is it that in OpenGL the method I currently use is perfectly possible but with modern API like Metal or maybe Vulkan too or using Sokol, you can only have fix buffers sizes and they can only be updated once? 
That is what I don't get. How for an example would you draw multiple 3D objects with higher vertex count? Lets say you draw 200 dragons on the screen like you do in one Oryol example. Lets say each model would have 1000 quads x 4 vertices. With 200 models, that alone would be around 3 GB if I calculated it right. Clearly you do some vertex animation here. So the positions of each vertex change each frame. And yet it seems possible. Guess there is some vodoo magic going on which I currently do not understand. 

If you say that would be definitely possible with Sokol, I will continue to study and hopefully understand what I have to do. I am slowly getting the hang of it and would love to continue.

Again, thank you for you time
Michael

**Comment by floooh** [comment]
If you can merge the images into one (aka a texture or sprite atlas) then you could render those 50k images at the correct Z order efficiently in a single buffer update and draw call. But even if you can't merge the images into one, you can still merge rectangles using the same image into a single draw call, this should cut down draw calls dramatically in most cases (that's what sokol_gl.h is doing (https://github.com/floooh/sokol/blob/5fbb6a25501e5478674ca0600a7539033592d749/util/sokol_gl.h#L400-L418).

Buffer size shouldn't affect rendering performance, putting hundreds of thousands of vertices into a single buffer isn't a problem, and updating those with a *single* sg_update_buffer() call also isn't.

> My last question....

Traditional APIs like OpenGL or D3D11 have a lot of internal buffer-management complexity to hide the details of dynamic data updates under a simple API in order to prevent the CPU having to wait for the GPU. This has upsides (easy usage if the update pattern is what GL and D3D have anticipated), and downsides (more complexity in the driver layer, not enough control if the update pattern isn't quite what GL or D3D expect).

The modern 3D APIs don't have this internal complexity, and move the burden of dynamic data updated to the API user. This gives the API user more control, but they need to implement their own dynamic data management.

The Oryol Dragon example is fairly clever in how it reduces the dynamic data that must be computed on th CPU and moved to the GPU. Each Dragon is a skinned character with a simple skeleton inside (you can see the skeleton in this demo: http://floooh.github.io/oryol-samples/wasm/OrbViewer.html, switch off "draw mesh" and switch on "draw animated pose". This simple skeleton is what's computed on the CPU for each Dragon, so instead of computing thousands of vertices, only a few dozen skeleton vertices per Dragon are computed. Then the skeleton data is written into a dynamic texture and uploaded to the GPU with a single update call (you can see that "Bone Texture" here by clicking the checkbox: http://floooh.github.io/oryol-samples/wasm/Dragons.html).

The dragons are then rendered via hardware instancing, which means the entire scene is rendering with a single draw clal no matter how many Dragons are rendered. The vertex shader reads the "Bone Texture" and computes the animated skin vertex positions. So all the hard work (updating the millions of dynamic vertices) is happening on the GPU alone.

The CPU only updates a few thousand bone vertices per frame and only does a single dynamic data upload each frame.

**Comment by MikeHart66** [comment]
Thank you Florian for the great explanation, especially about how the dragon example is managed. That gave me some ideas. Need to try and see if that could work.
